compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -3488,7 +3488,7 @@ class MatchReducer(initctx: Context) extends TypeComparer(initctx) {
                 false
 
           case MatchTypeCasePattern.AbstractTypeConstructor(tycon, argPatterns) =>
-            scrut.dealias match
+            scrut.dealiasKeepOpaques match
```

Comment:
This change is direct contradiction with the spec, and will break as many things as it will fix (since another use case might be to precisely take private knowledge into account to be able to perform a reduction).

Reply:
If I understood correctly, the following
```scala 3
object OpaqueScope:
  opaque type Opaque[N <: Any] = Double
  val foo: StripOpaque[Opaque[Int]] = 1

  // Note that match types within OpaqueScope are not limited by the absence of dealiasing
  type NeedsRhs[D] = D match
    case Double => String
  val bar: NeedsRhs[Opaque[Float]] = "hello"
```
is an example of what you are mentioning right ?

Reply:
Yes, exactly.

Reply:
Then it still works, and is part of the test.

Reply:
Oh yes because you only changed the dealiasing rule for the `AbstractTypeConstructor` case. You have to complicate the test a bit more so that it would also fall into that case to observe the issue.